Building quality apps/Unit tests.md

```diff
@@ -0,0 +1,155 @@
+* Fake - An object that has a working implementation, but not same as production one. Usually it takes some shortcut and has a simplified version of the production code. For example faking a database with an in memory solution in order to make our unit tests faster.
+
+
+When we talk about mocks and stubs we do not need to write our own implementation of those test doubles. Luckily, we have a tool called Mockito which enables us to mock and stub behaviours and dependencies with ease. Keep in mind that Mockito does not enable us to create fakes.
+
+Let's take a look at two examples where we use Mockito stubbing and mocking features.
+
+* Using Mockito for mocking
+
+```kotlin  
+val authTokenManager = mock(AuthTokenManager::class.java)
+verify(authTokenManger).saveAuthToken("Token")
+```
+With mocks we are verifying a specific behavior which can be seen in the above example using the verify method from Mockito.
+
+* Using Mockito for stubbing
+
+```kotlin  
+val authTokenManager = mock(AuthTokenManager::class.java)
+whenever(authTokenManager.getAuthToken()).thenReturn("Token")
+```
+With stubs we predefine results for a specific method so that just like it is shown in the above example. Every time the `getAuthToken` method is called we will get a stubbed result which is in our case `"Token"`.
+
+
+It is important to know this terminology because it is essentially part of the  testing fundamentals and it will definitely help you on your journey in becoming better in writing tests. You will also often find these terms in documentation of some testing library that you will use.
+
+### Write testable code
+
+This one is easier said than done. It is a good practice to actually think about your tests when writing production code. In general it is easy to get a picture how you will test something. If you ever get stuck and you are not sure how to test something that you actually wrote, then it is a right moment to consult a colleague or research the problem in more detail. 
+
+Some general guideline when writing testable code is to:
+
+* **Use dependancy injection** - you will hear this a lot but one can not stress enough how important this is. This approach allows us to write loosely coupled code which, in turn, enables us to replace production dependencies with appropriate test doubles. As an end result this allows us to isolate units of work and write good unit tests.
+* **Use abstraction** - it is a good rule of thumb to abstract an implementation in case you are not sure if you can easily test it or not. For example, the use of abstraction allows us to hide a concrete class that uses android specific implementation. Since android is not available on JVM it makes sense to hide it behind an abstraction which we can easily mock in our tests. 
```

Comment:
This is great, however I would add another thing and that is to avoid android code in tests. yes, you mention i
t in the last sentence, but in my opinion, it's not clear enough that we shouldn't have android related stuff in unit tests.

Reply:
I will see what we can do about it, maybe the changes in the beginning that tepi suggested could shed some more light on this issue